lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php

```diff
@@ -683,10 +690,37 @@ protected function prepareUpdateData($entity, bool $isInsert = false)
             if ($newVal !== null) {
                 $oid = spl_object_id($newVal);
 
-                if (isset($this->queuedInserts[$oid]) || $uow->isScheduledForInsert($newVal)) {
-                    // The associated entity $newVal is not yet persisted, so we must
-                    // set $newVal = null, in order to insert a null value and schedule an
-                    // extra update on the UnitOfWork.
+                // If the associated entity $newVal is not yet persisted and/or does not yet have
+                // an ID assigned, we must set $newVal = null. This will insert a null value and
+                // schedule an extra update on the UnitOfWork.
+                //
+                // This gives us extra time to a) possibly obtain a database-generated identifier
+                // value for $newVal, and b) insert $newVal into the database before the foreign
+                // key reference is being made.
+                //
+                // Depending on the ID generator strategy used, which entity persister is responsible
+                // for inserting the entity and whether the UoW already had an opportunity to write
+                // database-generated IDs ("post insert id") back into the entities, combinations of
+                // these factors may occur. For example: An entity may have an application-provided (natural)
+                // ID, but has not yet been inserted into the database. Or, we used the "AUTO" generator
+                // strategy and have already inserted the entity into the database, but the ID is not yet
+                // available in the entity's fields (it's set by the UoW after the EntityPersister finished).
+                //
+                // When looking at $this->queuedInserts and $uow->isScheduledForInsert, be aware
+                // of the implementation details that our own executeInserts() method will remove
+                // entities from the former as soon as the insert statement has been executed, and
+                // that the UnitOfWork has already removed entities from its own list at the time
+                // they were passed to our addInsert() method.
+                //
+                // Then, there is one extra exception we can make: An entity that references back to itself
+                // _and_ uses an application-provided ID (the "NONE" generator strategy) also does not
+                // need the extra update, although it is still in the list of insertions itself.
+                // This looks like a minor optimization at first, but is the capstone for being able to
+                // use non-NULLable, self-referencing associations in applications that provide IDs (like UUIDs).
+                if (
+                    (isset($this->queuedInserts[$oid]) || $uow->isScheduledForInsert($newVal) || !$uow->isInIdentityMap($newVal))
```

Comment:
This has to check the identity map once for every association value during inserts. 

Performance issue?